### PR TITLE
Add host_tx_ready enhancements

### PR DESCRIPTION
**Pull request by noaOrMlnx**
**What I did**
Add host_tx_ready enhancements to sonic-swss branch. related to HLD and commits: TODO add

**Why I did it**
Module and Asic configurations must be synchronized in order to not configure an non-existing module.

**How I verified it**
Manually

**Details if related**

### CODE REVIEW

File: orchagent/orchdaemon.cpp [hunk 1]
```diff
@@ -134,13 +134,27 @@ bool OrchDaemon::init()
         { APP_LAG_MEMBER_TABLE_NAME,  portsorch_base_pri     }
     };
 
+    vector<table_name_with_pri_t> state_transceiver_tables = {
+        {STATE_TRANSCEIVER_INFO_TABLE_NAME, portsorch_base_pri}
+    };
+
+    // PriTablesConnector appPriTables(m_applDb, ports_tables);
```
Comment (Junchao-Mellanox) [inline orchagent/orchdaemon.cpp]: Please remove unused comment

File: orchagent/orch.h [hunk 2]
```diff
@@ -214,6 +214,7 @@ typedef enum
 
 typedef std::pair<swss::DBConnector *, std::string> TableConnector;
 typedef std::pair<swss::DBConnector *, std::vector<std::string>> TablesConnector;
+typedef std::pair<swss::DBConnector *, const std::vector<table_name_with_pri_t> &> PriTablesConnector;
```
Comment (Junchao-Mellanox) [inline orchagent/orch.h]: Please remove unused change

File: orchagent/orch.h [hunk 3]
```diff
@@ -222,6 +223,8 @@ class Orch
     Orch(swss::DBConnector *db, const std::vector<std::string> &tableNames);
     Orch(swss::DBConnector *db, const std::vector<table_name_with_pri_t> &tableNameWithPri);
     Orch(const std::vector<TableConnector>& tables);
+    Orch(const std::vector<PriTablesConnector> &dbsTablesWithPri);
```
Comment (Junchao-Mellanox) [inline orchagent/orch.h]: Please remove unused change

File: orchagent/orchdaemon.cpp [hunk 4]
```diff
@@ -134,13 +134,27 @@ bool OrchDaemon::init()
         { APP_LAG_MEMBER_TABLE_NAME,  portsorch_base_pri     }
     };
 
+    vector<table_name_with_pri_t> state_transceiver_tables = {
```
Comment (Junchao-Mellanox) [inline orchagent/orchdaemon.cpp]: Pease remove unused variable

File: orchagent/orchdaemon.cpp [hunk 5]
```diff
@@ -134,13 +134,27 @@ bool OrchDaemon::init()
...
+    // PriTablesConnector appPriTables(m_applDb, ports_tables);
+    // PriTablesConnector statePriTables(m_stateDb, state_transceiver_tables);
+
+    // vector<PriTablesConnector> portsOrchTables = {
+    //     appPriTables,
+    //     statePriTables
+    // };
+
+    gPortsOrch = new PortsOrch(m_applDb, m_stateDb, ports_tables, m_chassisAppDb);
+    // gPortsOrch = new PortsOrch(m_applDb, m_stateDb, portsOrchTables, m_chassisAppDb);
```
Comment (Junchao-Mellanox) [inline orchagent/orchdaemon.cpp]: Please remove unused comments

File: orchagent/portsorch.cpp [hunk 6]
```diff
@@ -358,6 +358,10 @@ static void getPortSerdesAttr(PortSerdesAttrMap_t &map, const PortConfig &port)
  *    bridge. By design, SONiC switch starts with all bridge ports removed from
  *    default VLAN and all ports removed from .1Q bridge.
  */
+
+
+// PortsOrch::PortsOrch(DBConnector *db, DBConnector *stateDb, vector<PriTablesConnector> &tables, DBConnector *chassisAppDb) :
```
Comment (Junchao-Mellanox) [inline orchagent/portsorch.cpp]: Please remove unused comment

File: orchagent/portsorch.cpp [hunk 7]
```diff
@@ -498,6 +502,39 @@ PortsOrch::PortsOrch(DBConnector *db, DBConnector *stateDb, vector<table_name_wi
...
+            SWSS_LOG_DEBUG("SAI_SWITCH_ATTR_PORT_HOST_TX_READY_NOTIFY is true");
+            m_saiTxReadyNotifySupported = true;
+        }
+    }
+
+    if (m_saiHwTxSignalSupported && m_saiTxReadyNotifySupported)
+    {
+        SWSS_LOG_DEBUG("m_cmisModuleAsicSyncSupported is true");
+        m_cmisModuleAsicSyncSupported = true;
+        auto transceiverInfoTableName = STATE_TRANSCEIVER_INFO_TABLE_NAME;
```
Comment (Junchao-Mellanox) [inline orchagent/portsorch.cpp]: this Variable is not needed, please remove it and use STATE_TRANSCEIVER_INFO_TABLE_NAME directly

File: orchagent/portsorch.cpp [hunk 8]
```diff
@@ -4363,6 +4452,69 @@ void PortsOrch::doVlanMemberTask(Consumer &consumer)
...
+            SWSS_LOG_DEBUG("TRANSCEIVER_INFO table has changed - DEL command for port %s", alias.c_str());
+            
+            Port p;
+            if (m_pluggedModulesPort.find(alias) != m_pluggedModulesPort.end())
+            {
+                p = m_pluggedModulesPort[alias];
+                m_pluggedModulesPort.erase(alias);
+            }
+            SWSS_LOG_DEBUG("Setting host_tx_signal NOT allow for port %s", alias.c_str());
+            setSaiHostTxSignal(p, false);
```
Comment (Junchao-Mellanox) [inline orchagent/portsorch.cpp]: should we move this line to the if branch, or Port p might be not initialized

File: orchagent/portsorch.cpp [hunk 9]
```diff
@@ -4363,6 +4452,69 @@ void PortsOrch::doVlanMemberTask(Consumer &consumer)
...
+bool PortsOrch::setSaiHostTxSignal(Port port, bool enable)
+{
+    sai_attribute_t attr;
+    attr.id = SAI_PORT_ATTR_HOST_TX_SIGNAL_ENABLE;
+    attr.value.booldata = enable;
+    sai_status_t status = sai_port_api->set_port_attribute(port.m_port_id, &attr);
+
+    if (status != SAI_STATUS_SUCCESS)
+    {
+        SWSS_LOG_ERROR("Could not set port {} attribute {}");
```
Comment (Junchao-Mellanox) [inline orchagent/portsorch.cpp]: what is the "{}" used for? Do you mean "%d"? And please add the format argument.

File: orchagent/portsorch.cpp [hunk 10]
```diff
@@ -4761,7 +4913,11 @@ void PortsOrch::doTask(Consumer &consumer)
 
     string table_name = consumer.getTableName();
 
-    if (table_name == APP_PORT_TABLE_NAME)
+    if (table_name == STATE_TRANSCEIVER_INFO_TABLE_NAME && m_cmisModuleAsicSyncSupported)
```
Comment (Junchao-Mellanox) [inline orchagent/portsorch.cpp]: Do we need check `m_cmisModuleAsicSyncSupported` here? We will only subscribe transceiver table if m_cmisModuleAsicSyncSupported is true.

File: orchagent/portsorch.cpp [hunk 11]
```diff
@@ -5067,6 +5224,25 @@ bool PortsOrch::initializePort(Port &port)
...
+    if (m_cmisModuleAsicSyncSupported)
+    {
+        bool hostTxReadyVal;
+        if (!getPortHostTxReady(port, hostTxReadyVal))
+        {
+            SWSS_LOG_ERROR("Failed to get host_tx_ready value from SAI to Port %" PRIx64 , port.m_port_id);
+        }
+        /* set value to state DB */
+        
+        string hostTxReadyStr = "false";
```
Comment (Junchao-Mellanox) [inline orchagent/portsorch.cpp]: suggest:
```
string hostTxReadyStr = hostTxReadyVal ? "true" : "false";
```

File: orchagent/portsorch.cpp [hunk 12]
```diff
@@ -1432,6 +1503,24 @@ bool PortsOrch::getPortAdminStatus(sai_object_id_t id, bool &up)
...
+bool PortsOrch::getPortHostTxReady(const Port& port, bool &hostTxReadyVal)
+{
+    SWSS_LOG_ENTER();
+
+    sai_attribute_t attr;
+    attr.id = SAI_PORT_ATTR_HOST_TX_READY_STATUS;
+
+    sai_status_t status = sai_port_api->get_port_attribute(port.m_port_id, 1, &attr);
+    if (status != SAI_STATUS_SUCCESS)
+    {
```
Comment (Junchao-Mellanox) [inline orchagent/portsorch.cpp:1551]: do we need set hostTxReadyVal to false here? I suppose we don't want leave it uninitialized.

File: orchagent/portsorch.cpp [hunk 13]
```diff
@@ -7177,6 +7353,27 @@ void PortsOrch::doTask(NotificationConsumer &consumer)
...
+    else if (&consumer == m_portHostTxReadyNotificationConsumer && op == "port_host_tx_ready")
+    {        
+        sai_object_id_t port_id;
+        sai_object_id_t switch_id;
+        sai_port_host_tx_ready_status_t host_tx_ready_status;
+
+        sai_deserialize_port_host_tx_ready_ntf(data, switch_id, port_id, host_tx_ready_status);
+        SWSS_LOG_DEBUG("Recieved host_tx_ready notification for port 0x%" PRIx64, port_id);
+
+        if (host_tx_ready_status == SAI_PORT_HOST_TX_READY_STATUS_READY)
```
Comment (Junchao-Mellanox) [inline orchagent/portsorch.cpp]: suggest:
```
setHostTxReady(port_id, host_tx_ready_status == SAI_PORT_HOST_TX_READY_STATUS_READY ? "true" : "false");
```

File: orchagent/portsorch.h [hunk 14]
```diff
@@ -120,6 +120,7 @@ class PortsOrch : public Orch, public Subject
 {
 public:
     PortsOrch(DBConnector *db, DBConnector *stateDb, vector<table_name_with_pri_t> &tableNames, DBConnector *chassisAppDb);
+    // PortsOrch(DBConnector *db, DBConnector *stateDb, vector<PriTablesConnector> &tables, DBConnector *chassisAppDb);
```
Comment (Junchao-Mellanox) [inline orchagent/portsorch.h]: please remove unused comment

File: orchagent/portsorch.h [hunk 15]
```diff
@@ -323,18 +325,25 @@ class PortsOrch : public Orch, public Subject
...
 
     NotificationConsumer* m_portStatusNotificationConsumer;
+    NotificationConsumer* m_portHostTxReadyNotificationConsumer;
+
     bool fec_override_sup = false;
     bool oper_fec_sup = false;
 
     swss::SelectableTimer *m_port_state_poller = nullptr;
 
+    bool m_saiHwTxSignalSupported = false;
```
Comment (Junchao-Mellanox) [inline orchagent/portsorch.h]: do you really need this member?

File: orchagent/portsorch.h [hunk 16]
```diff
@@ -323,18 +325,25 @@ class PortsOrch : public Orch, public Subject
...
+    bool m_saiTxReadyNotifySupported = false;
```
Comment (Junchao-Mellanox) [inline orchagent/portsorch.h]: do you really need this member?

File: orchagent/portsorch.h [hunk 17]
```diff
@@ -388,6 +398,9 @@ class PortsOrch : public Orch, public Subject
 
     bool setBridgePortAdminStatus(sai_object_id_t id, bool up);
 
+    bool setSaiHostTxSignal(Port port, bool enable);
```
Comment (Junchao-Mellanox) [inline orchagent/portsorch.h]: const Port& port?

File: orchagent/portsorch.h [hunk 18]
```diff
@@ -388,6 +398,9 @@ class PortsOrch : public Orch, public Subject
...
+
+    void setHostTxReady(sai_object_id_t portId, std::string status);
```
Comment (Junchao-Mellanox) [inline orchagent/portsorch.h]: const string& status?

File: orchagent/portsorch.cpp [hunk 19]
```diff
@@ -1380,7 +1434,10 @@ bool PortsOrch::setPortAdminStatus(Port &port, bool state)
         SWSS_LOG_ERROR("Failed to set admin status %s for port %s."
                        " Setting host_tx_ready as false",
                        state ? "UP" : "DOWN", port.m_alias.c_str());
-        m_portStateTable.hset(port.m_alias, "host_tx_ready", "false");
+        if (!m_cmisModuleAsicSyncSupported)
```
Comment (Junchao-Mellanox) [inline orchagent/portsorch.cpp:1473]: do you still need this if?

**Review by Junchao-Mellanox — CHANGES_REQUESTED** [review]

File: orchagent/portsorch.cpp (hunk 19 above)
Comment (noaOrMlnx) [reply orchagent/portsorch.cpp:1473]: why would I not need it? 
the purpose of it is to set host_tx_ready=false by default only when we work in dependent mode.

File: orchagent/orchdaemon.cpp (hunk 1 above)
Comment (noaOrMlnx) [reply orchagent/orchdaemon.cpp]: Done

File: orchagent/orch.h (hunk 2 above)
Comment (noaOrMlnx) [reply orchagent/orch.h]: Done

File: orchagent/orchdaemon.cpp (hunk 4 above)
Comment (noaOrMlnx) [reply orchagent/orchdaemon.cpp]: Done

File: orchagent/portsorch.h (hunk 15 above)
Comment (noaOrMlnx) [reply orchagent/portsorch.h]: Done

File: orchagent/portsorch.h (hunk 16 above)
Comment (noaOrMlnx) [reply orchagent/portsorch.h]: Done

File: orchagent/orchdaemon.cpp (hunk 5 above)
Comment (noaOrMlnx) [reply orchagent/orchdaemon.cpp]: Done

File: orchagent/orch.h (hunk 3 above)
Comment (noaOrMlnx) [reply orchagent/orch.h]: Done

File: orchagent/portsorch.cpp (hunk 6 above)
Comment (noaOrMlnx) [reply orchagent/portsorch.cpp]: Done

File: orchagent/portsorch.cpp (hunk 8 above)
Comment (noaOrMlnx) [reply orchagent/portsorch.cpp]: Done

File: orchagent/portsorch.cpp (hunk 9 above)
Comment (noaOrMlnx) [reply orchagent/portsorch.cpp]: Done

File: orchagent/portsorch.cpp (hunk 11 above)
Comment (noaOrMlnx) [reply orchagent/portsorch.cpp]: Done

File: orchagent/portsorch.cpp (hunk 13 above)
Comment (noaOrMlnx) [reply orchagent/portsorch.cpp]: Done

File: orchagent/orch.h [hunk 20]
```diff
@@ -222,6 +222,7 @@ class Orch
     Orch(swss::DBConnector *db, const std::vector<std::string> &tableNames);
     Orch(swss::DBConnector *db, const std::vector<table_name_with_pri_t> &tableNameWithPri);
     Orch(const std::vector<TableConnector>& tables);
+
```
Comment (Junchao-Mellanox) [inline orchagent/orch.h]: it seems you don't change this file, please exclude this file from your PR.

File: orchagent/orchdaemon.cpp [hunk 21]
```diff
@@ -134,13 +134,14 @@ bool OrchDaemon::init()
         { APP_LAG_MEMBER_TABLE_NAME,  portsorch_base_pri     }
     };
 
+    gPortsOrch = new PortsOrch(m_applDb, m_stateDb, ports_tables, m_chassisAppDb);
```
Comment (Junchao-Mellanox) [inline orchagent/orchdaemon.cpp]: Do you have actual changes to this file? If not, please exclude it from your PR.

**Pull request closed without merging**